### PR TITLE
Multi-file Swift Output

### PR DESCRIPTION
**Pull request by MrRooni**
## Reason to Be

When running quicktype to generate Swift code from a directory of schema files the output currently dumps everything into a single `.swift` file. This PR breaks each named type out into its own file. Additionally it generates a `JSONSchemaSupport.swift` that contains many of the helper functions needed by the various outputs.

Resolves https://github.com/quicktype/quicktype/issues/687

## CHANGELOG Entry

> [IMPROVED] Generating Swift files from a directory of inputs now outputs one Swift file per top level schema object.

## Thought Process

After seeing that the Java renderer already does this exact thing and digging into how quicktype writes its files I began refactoring the renderer to divide up the generated source into sensible blocks.

## How To Test

Testing manually should be as simple as:
```
quicktype . -l Swift -s schema -o .
```

## Possible Impacts

I know for a fact that I've broken the rendering of one of the schema inputs. _shakes fist at `keyword-unions.schema`_

I'll be working to get that fixed up…somehow.

### CODE REVIEW

**Comment by MrRooni** [comment]
Throwing an informal `wait` tag on this one. Need to tidy up a couple things first, but wanted to see how the tests turn out from the server.

**Comment by schani** [comment]
Looks like the emitted code is incorrect:

```
main.swift:12:28: error: value of optional type 'TopLevel?' not unwrapped; did you mean to use '!' or '?'?
  | guard let jsonString = try obj.jsonString() else {
  | ^
  | ?
  |  
  | Fixture threw an exception
  | {
  | "error": {
  | "command": "swiftc -o quicktype main.swift quicktype.swift",
  | "code": 1
  | },
  | "sample": {
  | "path": "test/inputs/json/priority/combinations1.json",
  | "additionalRendererOptions": {
  | "support-linux": "false"
  | },
  | "saveOutput": false
  | },
  | "cwd": "/app/test/runs/swift-4f7779"
  | }
  |  
  | main.swift:12:28: error: value of optional type 'TopLevel?' not unwrapped; did you mean to use '!' or '?'?
  | guard let jsonString = try obj.jsonString() else {
  | ^
  | ?
  |  
  | Fixture threw an exception
  | {
  | "error": {
  | "command": "swiftc -o quicktype main.swift quicktype.swift",
  | "code": 1
  | },
  | "sample": {
  | "path": "test/inputs/json/priority/combinations2.json",
  | "additionalRendererOptions": {
  | "support-linux": "false"
  | },
  | "saveOutput": false
  | },
  | "cwd": "/app/test/runs/swift-031586"
  | }
  |  
  | main.swift:12:28: error: value of optional type 'TopLevel?' not unwrapped; did you mean to use '!' or '?'?
  | guard let jsonString = try obj.jsonString() else {
  | ^
  | ?
  |  
  | Fixture threw an exception
  | {
  | "error": {
  | "command": "swiftc -o quicktype main.swift quicktype.swift",
  | "code": 1
  | },
  | "sample": {
  | "path": "test/inputs/json/priority/combinations3.json",
  | "additionalRendererOptions": {
  | "support-linux": "false"
  | },
  | "saveOutput": false
  | },
  | "cwd": "/app/test/runs/swift-2cfdc9"
  | }
```

**Comment by schani** [comment]
```
JSONSchemaSupport.swift:33:41: error: use of undeclared type 'Hasher'
  | public func hash(into hasher: inout Hasher) {
  | ^~~~~~
  |  
  | Fixture threw an exception
  | {
  | "error": {
  | "command": "swiftc -o quicktype *.swift",
  | "code": 1
  | },
  | "sample": {
  | "path": "test/inputs/json/priority/combinations3.json",
  | "additionalRendererOptions": {
  | "support-linux": "false"
  | },
  | "saveOutput": false
  | },
  | "cwd": "/app/test/runs/swift-30719e"
  | }
  |  
  | JSONSchemaSupport.swift:33:41: error: use of undeclared type 'Hasher'
  | public func hash(into hasher: inout Hasher) {
  | ^~~~~~
  |  
  | Fixture threw an exception
  | {
  | "error": {
  | "command": "swiftc -o quicktype *.swift",
  | "code": 1
  | },
  | "sample": {
  | "path": "test/inputs/json/priority/combinations4.json",
  | "additionalRendererOptions": {
  | "support-linux": "false"
  | },
  | "saveOutput": false
  | },
  | "cwd": "/app/test/runs/swift-8ba570"
  | }
```

**Comment by schani** [comment]
```
JSONSchemaSupport.swift:33:41: error: use of undeclared type 'Hasher'
  | public func hash(into hasher: inout Hasher) {
  | ^~~~~~
  |  
  | Fixture threw an exception
  | {
  | "error": {
  | "command": "swiftc -o quicktype *.swift",
  | "code": 1
  | },
  | "sample": {
  | "path": "test/inputs/json/priority/combinations3.json",
  | "additionalRendererOptions": {
  | "support-linux": "false"
  | },
  | "saveOutput": false
  | },
  | "cwd": "/app/test/runs/swift-511bfd"
  | }
```

File: build/quicktype-core/package.in.json [hunk 1]
```diff
@@ -20,15 +20,13 @@
         "pluralize": "^7.0.0",
         "@mark.probst/unicode-properties": "~1.1.0",
         "urijs": "^1.19.1",
-        "wordwrap": "^1.0.0",
-        "yaml": "^1.5.0"
+        "wordwrap": "^1.0.0"
```
Comment (schani) [inline build/quicktype-core/package.in.json]: Why is YAML removed here?

File: src/quicktype-core/Renderer.ts [hunk 2]
```diff
@@ -266,10 +290,20 @@ export abstract class Renderer {
...
+                this._emitContext = existingEmitContext;
+            }
+        }
+    }
+
     protected finishFile(filename: string): void {
         assert(!this._finishedFiles.has(filename), `Tried to emit file ${filename} more than once`);
         const source = sourcelikeToSource(this._emitContext.source);
         this._finishedFiles.set(filename, source);
+        this._finishedEmitContexts.set(filename.toLowerCase().toString(), this._emitContext);
```
Comment (schani) [inline src/quicktype-core/Renderer.ts]: This looks weird.  Why is the filename lowercased, and why is `toString()` called on it?

File: src/quicktype-core/Renderer.ts [hunk 3]
```diff
@@ -266,10 +290,20 @@ export abstract class Renderer {
         return assignNames(this.setUpNaming());
     }
 
+    protected initializeEmitContextForFilename(filename: string): void {
+        if (this._finishedEmitContexts.has(filename.toLowerCase().toString())) {
+            const existingEmitContext = this._finishedEmitContexts.get(filename.toLowerCase().toString());
```
Comment (schani) [inline src/quicktype-core/Renderer.ts]: `.toLowerCase().toString()`?  Is the `toLowerCase()` to deal with case-insensitive file systems?  Will the code work with case-sensitive file systems in all cases, too?

File: build/quicktype-core/package.in.json (hunk 1 above)
Comment (MrRooni) [reply build/quicktype-core/package.in.json]: It must have been changed and mistakenly committed when I was getting my build and test environment up and running. I've put it back.

File: src/quicktype-core/Renderer.ts (hunk 2 above)
Comment (MrRooni) [reply src/quicktype-core/Renderer.ts]: Because I mistakenly thought `toLowerCase()` returned a `String` not a `string`. Fixed.

**Comment by schani** [comment]
```
JSONSchemaSupport.swift:33:41: error: use of undeclared type 'Hasher'
  | public func hash(into hasher: inout Hasher) {
  | ^~~~~~
  |  
  | Fixture threw an exception
  | {
  | "error": {
  | "command": "swiftc -o quicktype *.swift",
  | "code": 1
  | },
  | "sample": {
  | "path": "test/inputs/json/priority/combinations4.json",
  | "additionalRendererOptions": {
  | "support-linux": "false"
  | },
  | "saveOutput": false
  | },
  | "cwd": "/app/test/runs/swift-c28655"
  | }
  | [object Object]
  | at /app/test/test.ts:68:17
  | at step (/app/test/test.ts:32:23)
  | at Object.throw (/app/test/test.ts:13:53)
  | at rejected (/app/test/test.ts:5:65)
  | at <anonymous>
  | at process._tickDomainCallback (internal/process/next_tick.js:229:7)
  |  
  | JSONSchemaSupport.swift:33:41: error: use of undeclared type 'Hasher'
  | public func hash(into hasher: inout Hasher) {
  | ^~~~~~
  |  
  | Fixture threw an exception
  | {
  | "error": {
  | "command": "swiftc -o quicktype *.swift",
  | "code": 1
  | },
  | "sample": {
  | "path": "test/inputs/json/priority/combinations3.json",
  | "additionalRendererOptions": {
  | "support-linux": "false"
  | },
  | "saveOutput": false
  | },
  | "cwd": "/app/test/runs/swift-a4533e"
  | }
```

**Comment by MrRooni** [comment]
@schani I should be able to make single file output the default, yes. I'll get that fixed up. Thanks for posting those errors. I'll be interested in seeing how the latest push works, those tests all pass locally for me.

File: src/quicktype-core/Renderer.ts (hunk 3 above)
Comment (MrRooni) [reply src/quicktype-core/Renderer.ts]: Yes, it will.

**Comment by schani** [comment]
```
quicktype.swift:339:22: error: invalid redeclaration of 'codableTask(with:completionHandler:)'
  | fileprivate func codableTask<T: Codable>(with url: URL, completionHandler: @escaping (T?, URLResponse?, Error?) -> Void) -> URLSessionDataTask {
  | ^
  | quicktype.swift:184:22: note: 'codableTask(with:completionHandler:)' previously declared here
  | fileprivate func codableTask<T: Codable>(with url: URL, completionHandler: @escaping (T?, URLResponse?, Error?) -> Void) -> URLSessionDataTask {
  | ^
  | quicktype.swift:505:22: error: invalid redeclaration of 'codableTask(with:completionHandler:)'
  | fileprivate func codableTask<T: Codable>(with url: URL, completionHandler: @escaping (T?, URLResponse?, Error?) -> Void) -> URLSessionDataTask {
  | ^
  | quicktype.swift:184:22: note: 'codableTask(with:completionHandler:)' previously declared here
  | fileprivate func codableTask<T: Codable>(with url: URL, completionHandler: @escaping (T?, URLResponse?, Error?) -> Void) -> URLSessionDataTask {
  | ^
  | quicktype.swift:630:22: error: invalid redeclaration of 'codableTask(with:completionHandler:)'
  | fileprivate func codableTask<T: Codable>(with url: URL, completionHandler: @escaping (T?, URLResponse?, Error?) -> Void) -> URLSessionDataTask {
  | ^
  | quicktype.swift:184:22: note: 'codableTask(with:completionHandler:)' previously declared here
  | fileprivate func codableTask<T: Codable>(with url: URL, completionHandler: @escaping (T?, URLResponse?, Error?) -> Void) -> URLSessionDataTask {
  | ^
  | quicktype.swift:831:22: error: invalid redeclaration of 'codableTask(with:completionHandler:)'
  | fileprivate func codableTask<T: Codable>(with url: URL, completionHandler: @escaping (T?, URLResponse?, Error?) -> Void) -> URLSessionDataTask {
  | ^
  | quicktype.swift:184:22: note: 'codableTask(with:completionHandler:)' previously declared here
  | fileprivate func codableTask<T: Codable>(with url: URL, completionHandler: @escaping (T?, URLResponse?, Error?) -> Void) -> URLSessionDataTask {
  | ^
  | quicktype.swift:1077:22: error: invalid redeclaration of 'codableTask(with:completionHandler:)'
  | fileprivate func codableTask<T: Codable>(with url: URL, completionHandler: @escaping (T?, URLResponse?, Error?) -> Void) -> URLSessionDataTask {
  | ^
  | quicktype.swift:184:22: note: 'codableTask(with:completionHandler:)' previously declared here
  | fileprivate func codableTask<T: Codable>(with url: URL, completionHandler: @escaping (T?, URLResponse?, Error?) -> Void) -> URLSessionDataTask {
  | ^
  | quicktype.swift:1271:22: error: invalid redeclaration of 'codableTask(with:completionHandler:)'
  | fileprivate func codableTask<T: Codable>(with url: URL, completionHandler: @escaping (T?, URLResponse?, Error?) -> Void) -> URLSessionDataTask {
  | ^
  | quicktype.swift:184:22: note: 'codableTask(with:completionHandler:)' previously declared here
  | fileprivate func codableTask<T: Codable>(with url: URL, completionHandler: @escaping (T?, URLResponse?, Error?) -> Void) -> URLSessionDataTask {
  | ^
  | quicktype.swift:1425:22: error: invalid redeclaration of 'codableTask(with:completionHandler:)'
  | fileprivate func codableTask<T: Codable>(with url: URL, completionHandler: @escaping (T?, URLResponse?, Error?) -> Void) -> URLSessionDataTask {
  | ^
  | quicktype.swift:184:22: note: 'codableTask(with:completionHandler:)' previously declared here
  | fileprivate func codableTask<T: Codable>(with url: URL, completionHandler: @escaping (T?, URLResponse?, Error?) -> Void) -> URLSessionDataTask {
  | ^
  | quicktype.swift:1620:22: error: invalid redeclaration of 'codableTask(with:completionHandler:)'
  | fileprivate func codableTask<T: Codable>(with url: URL, completionHandler: @escaping (T?, URLResponse?, Error?) -> Void) -> URLSessionDataTask {
  | ^
  | quicktype.swift:184:22: note: 'codableTask(with:completionHandler:)' previously declared here
  | fileprivate func codableTask<T: Codable>(with url: URL, completionHandler: @escaping (T?, URLResponse?, Error?) -> Void) -> URLSessionDataTask {
  | ^
  | quicktype.swift:1866:22: error: invalid redeclaration of 'codableTask(with:completionHandler:)'
  | fileprivate func codableTask<T: Codable>(with url: URL, completionHandler: @escaping (T?, URLResponse?, Error?) -> Void) -> URLSessionDataTask {
  | ^
  | quicktype.swift:184:22: note: 'codableTask(with:completionHandler:)' previously declared here
  | fileprivate func codableTask<T: Codable>(with url: URL, completionHandler: @escaping (T?, URLResponse?, Error?) -> Void) -> URLSessionDataTask {
  | ^
  | quicktype.swift:2020:22: error: invalid redeclaration of 'codableTask(with:completionHandler:)'
  | fileprivate func codableTask<T: Codable>(with url: URL, completionHandler: @escaping (T?, URLResponse?, Error?) -> Void) -> URLSessionDataTask {
  | ^
  | quicktype.swift:184:22: note: 'codableTask(with:completionHandler:)' previously declared here
  | fileprivate func codableTask<T: Codable>(with url: URL, completionHandler: @escaping (T?, URLResponse?, Error?) -> Void) -> URLSessionDataTask {
  | ^
  | quicktype.swift:2179:22: error: invalid redeclaration of 'codableTask(with:completionHandler:)'
  | fileprivate func codableTask<T: Codable>(with url: URL, completionHandler: @escaping (T?, URLResponse?, Error?) -> Void) -> URLSessionDataTask {
  | ^
  | quicktype.swift:184:22: note: 'codableTask(with:completionHandler:)' previously declared here
  | fileprivate func codableTask<T: Codable>(with url: URL, completionHandler: @escaping (T?, URLResponse?, Error?) -> Void) -> URLSessionDataTask {
  | ^
  | quicktype.swift:2509:22: error: invalid redeclaration of 'codableTask(with:completionHandler:)'
  | fileprivate func codableTask<T: Codable>(with url: URL, completionHandler: @escaping (T?, URLResponse?, Error?) -> Void) -> URLSessionDataTask {
  | ^
  | quicktype.swift:184:22: note: 'codableTask(with:completionHandler:)' previously declared here
  | fileprivate func codableTask<T: Codable>(with url: URL, completionHandler: @escaping (T?, URLResponse?, Error?) -> Void) -> URLSessionDataTask {
  | ^
  | quicktype.swift:2808:22: error: invalid redeclaration of 'codableTask(with:completionHandler:)'
  | fileprivate func codableTask<T: Codable>(with url: URL, completionHandler: @escaping (T?, URLResponse?, Error?) -> Void) -> URLSessionDataTask {
  | ^
  | quicktype.swift:184:22: note: 'codableTask(with:completionHandler:)' previously declared here
  | fileprivate func codableTask<T: Codable>(with url: URL, completionHandler: @escaping (T?, URLResponse?, Error?) -> Void) -> URLSessionDataTask {
  | ^
  |  
  | Fixture threw an exception
  | {
  | "error": {
  | "command": "swiftc -o quicktype main.swift quicktype.swift",
  | "code": 1
  | },
  | "sample": {
  | "path": "test/inputs/json/priority/combinations1.json",
  | "additionalRendererOptions": {
  | "url-session": "true"
  | },
  | "saveOutput": false
  | },
  | "cwd": "/app/test/runs/swift-eb5a9b"
  | }
```

**Comment by schani** [comment]
```
Command failed
  | {
  | "command": "diff -Naur quicktype.swift.expected quicktype.swift > /dev/null 2>&1",
  | "code": 1,
  | "cwd": "/app/test/runs/swift-01855a"
  | }
  | Fixture threw an exception
  | {
  | "error": {
  | "command": "diff -Naur quicktype.swift.expected quicktype.swift > /dev/null 2>&1",
  | "code": 1,
  | "cwd": "/app/test/runs/swift-01855a"
  | },
  | "sample": {
  | "path": "test/inputs/json/priority/bug863.json",
  | "additionalRendererOptions": {},
  | "saveOutput": true
  | },
  | "cwd": "/app/test/runs/swift-01855a"
  | }
```

What failed here is another test we do: It generates JSON Schema from the input JSON, and then it generates Swift from the JSON Schema, and diffs that Swift output with the one coming directly from the input JSON.  Ideally those should be identical, but here's a case where they differ.  Could you look into that, please?  As you can see in `languages.ts` there's a long list of inputs in `skipDiffViaSchema` that we already exempt from that test because it's a bit brittle (in the case of Swift it's mostly issues with date/time).

**Comment by MrRooni** [comment]
I'm on it, thanks @schani

**Comment by schani** [comment]
[Here's the diff](https://s3.amazonaws.com/quicktype-outputs/diffs/4c086b1ca85b379db0e97cc6cf1b5b4e8e52e7ea-3248825acc7b54fcd3db27b41831be6c298066c5.diff).

One thing that jumps out right away is that the header

```
+// To parse the JSON, add this file to your project and do:
+//
+//   let dataClass = try DataClass(json)
```

is added multiple times per file.  That should be fixed.  I haven't looked at it in more detail apart from that.

**Comment by schani** [comment]
Awesome, thank you!

**Pull request merged**